File: Backend/src/main/java/lan/dk/podcastserver/manager/worker/finder/DailymotionFinder.java
package lan.dk.podcastserver.manager.worker.finder;

import com.fasterxml.jackson.annotation.JsonProperty;
import javaslang.control.Option;
import lan.dk.podcastserver.entity.Cover;
import lan.dk.podcastserver.entity.Podcast;
import lan.dk.podcastserver.service.ImageService;
import lan.dk.podcastserver.service.JsonService;
import lombok.Getter;
import lombok.RequiredArgsConstructor;
import lombok.Setter;
import org.hibernate.validator.constraints.NotEmpty;
import org.springframework.stereotype.Service;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

import static java.util.Objects.nonNull;

/**
 * Created by kevin on 23/02/2016 for Podcast Server
 */
@Service("DailymotionFinder")
@RequiredArgsConstructor
public class DailymotionFinder implements Finder {

    private static final String API_URL = "https://api.dailymotion.com/user/%s?fields=avatar_720_url,description,username";
    private static final Pattern USER_NAME_EXTRACTOR = Pattern.compile("^.+dailymotion.com\\/(.*)");

    final JsonService jsonService;
    final ImageService imageService;

    @Override
    public Podcast find(String url) {
        return usernameOf(url)
                .map(username -> String.format(API_URL, username))
                .flatMap(jsonService::parseUrl)
                .map(json -> json.read("$", DailymotionUserDetail.class))
                .map(d -> jsonToPodcast(url, d))
                .getOrElse(() -> Podcast.builder().url(url).type("Dailymotion").cover(Cover.DEFAULT_COVER).build());
    }

    private Podcast jsonToPodcast(String url, DailymotionUserDetail detail) {
        return Podcast.builder()
                .cover(imageService.getCoverFromURL(detail.getAvatar()))
                .url(url)
                .title(detail.getUsername())
                .description(detail.getDescription())
                .type("Dailymotion")
                .build();
    }

    private Option<String> usernameOf(String url) {
        // http://www.dailymotion.com/karimdebbache
        Matcher matcher = USER_NAME_EXTRACTOR.matcher(url);
        if (!matcher.find())
            return Option.none();

        return Option.of(matcher.group(1));
    }

    private static class DailymotionUserDetail {
        @JsonProperty("avatar_720_url") @Setter @Getter private String avatar;
        @Setter @Getter private String username;
        @Setter @Getter private String description;
    }

    public Integer compatibility(@NotEmpty String url) {
        return nonNull(url) && url.contains("www.dailymotion.com") ? 1 : Integer.MAX_VALUE;
    }

}

File: backend/src/main/kotlin/com/github/davinkevin/podcastserver/find/finders/dailymotion/DailymotionFinder.kt
package com.github.davinkevin.podcastserver.find.finders.dailymotion

import com.fasterxml.jackson.annotation.JsonIgnoreProperties
import com.fasterxml.jackson.annotation.JsonProperty
import com.github.davinkevin.podcastserver.find.FindPodcastInformation
import com.github.davinkevin.podcastserver.find.finders.Finder
import com.github.davinkevin.podcastserver.find.finders.fetchFindCoverInformation
import com.github.davinkevin.podcastserver.service.image.ImageService
import org.springframework.web.client.RestClient
import org.springframework.web.client.body
import reactor.core.publisher.Mono
import reactor.kotlin.core.publisher.toMono
import java.net.URI

/**
 * Created by kevin on 01/11/2019
 */
class DailymotionFinder(
    private val wc: RestClient,
    private val image: ImageService
): Finder {

    override fun findPodcastInformation(url: String): FindPodcastInformation? {
        val matches = USER_NAME_EXTRACTOR.find(url) ?: throw RuntimeException("username not found in url $url")
        val userName = matches.groups["id"]!!.value

        val detail = wc.get()
            .uri { it
                .pathSegment("user")
                .path(userName)
                .queryParam("fields", "avatar_720_url,description,username")
                .build()
            }
            .retrieve()
            .body<DailymotionUserDetail>()
            ?: return null

        val cover = detail.avatar
            .let(::URI)
            .let(image::fetchFindCoverInformation)

        return FindPodcastInformation(
            title = detail.username,
            url = URI(url),
            description = detail.description ?: "",
            type = "Dailymotion",
            cover = cover
        )
    }

    override fun findInformation(url: String): Mono<FindPodcastInformation> {
        return findPodcastInformation(url).toMono()
    }

    override fun compatibility(url: String): Int = when {
        "www.dailymotion.com" in url -> 1
        else -> Int.MAX_VALUE
    }
}

private val USER_NAME_EXTRACTOR = "^.+dailymotion.com/(?<id>.*)".toRegex()

@JsonIgnoreProperties(ignoreUnknown = true)
private class DailymotionUserDetail(
        @JsonProperty("avatar_720_url") val avatar: String,
        val username: String,
        val description: String?
)
